{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "d8875049",
   "metadata": {},
   "source": [
    "# __Extracting all Bifidobacterium strains__"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "602ca680",
   "metadata": {},
   "source": [
    "### Loading and cleaning the required datasets"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "26c25444",
   "metadata": {},
   "source": [
    "Import required libraries"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "505cc0eb",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import time\n",
    "from pathlib import Path\n",
    "from collections import defaultdict"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b2766b04",
   "metadata": {},
   "source": [
    "We will now load `names.dmp` and `nodes.dmp` as pandas DataFrames from the `new_taxdump_2025-12-01.zip` archive.  \n",
    "This allows us to work with the data in a clean and structured way.\n",
    "\n",
    "> Note: These files are not included in the repository due to their large size.  \n",
    "> You can download the `new_taxdump_2025-12-01` snapshot from the NCBI FTP site:  \n",
    "> [https://ftp.ncbi.nlm.nih.gov/pub/taxonomy/taxdump_archive/](https://ftp.ncbi.nlm.nih.gov/pub/taxonomy/taxdump_archive/)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "3f2d85ed",
   "metadata": {},
   "outputs": [],
   "source": [
    "names_path = Path('../data/raw/new_taxdump_2025-12-01/names.dmp')\n",
    "names = pd.read_csv(names_path,sep=r\"\\t\\|\\t\",engine='python',header=None)\n",
    "names.columns = ['tax_id','name_txt','unique_name','name_class']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "cb48c0b0",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>tax_id</th>\n",
       "      <th>name_txt</th>\n",
       "      <th>unique_name</th>\n",
       "      <th>name_class</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>all</td>\n",
       "      <td>NaN</td>\n",
       "      <td>synonym\\t|</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>root</td>\n",
       "      <td>NaN</td>\n",
       "      <td>scientific name\\t|</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Bacteria &lt;bacteria&gt;</td>\n",
       "      <td>scientific name\\t|</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>2</td>\n",
       "      <td>bacteria</td>\n",
       "      <td>bacteria &lt;blast name&gt;</td>\n",
       "      <td>blast name\\t|</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>2</td>\n",
       "      <td>bacteria</td>\n",
       "      <td>bacteria &lt;genbank common name&gt;</td>\n",
       "      <td>genbank common name\\t|</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   tax_id  name_txt                     unique_name              name_class\n",
       "0       1       all                             NaN              synonym\\t|\n",
       "1       1      root                             NaN      scientific name\\t|\n",
       "2       2  Bacteria             Bacteria <bacteria>      scientific name\\t|\n",
       "3       2  bacteria           bacteria <blast name>           blast name\\t|\n",
       "4       2  bacteria  bacteria <genbank common name>  genbank common name\\t|"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "names.head() "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "6f6bcbe7",
   "metadata": {},
   "outputs": [],
   "source": [
    "names['name_class'] = names['name_class'].str.strip('\\t|') # Remove unwanted characters from the 'name_class' column"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "38d4ae78",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>tax_id</th>\n",
       "      <th>name_txt</th>\n",
       "      <th>unique_name</th>\n",
       "      <th>name_class</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>all</td>\n",
       "      <td>NaN</td>\n",
       "      <td>synonym</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>root</td>\n",
       "      <td>NaN</td>\n",
       "      <td>scientific name</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Bacteria &lt;bacteria&gt;</td>\n",
       "      <td>scientific name</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   tax_id  name_txt          unique_name       name_class\n",
       "0       1       all                  NaN          synonym\n",
       "1       1      root                  NaN  scientific name\n",
       "2       2  Bacteria  Bacteria <bacteria>  scientific name"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "names.head(3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "2d120c05",
   "metadata": {},
   "outputs": [],
   "source": [
    "nodes_path = Path('../data/raw/new_taxdump_2025-12-01/nodes.dmp')\n",
    "nodes = pd.read_csv(nodes_path,engine='python',sep=r'\\t\\|\\t',header=None,usecols=[0,1,2],names=['tax_id','parent_tax_id','rank'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "6df01541",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>tax_id</th>\n",
       "      <th>parent_tax_id</th>\n",
       "      <th>rank</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>no rank</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>2</td>\n",
       "      <td>131567</td>\n",
       "      <td>domain</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>6</td>\n",
       "      <td>335928</td>\n",
       "      <td>genus</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>7</td>\n",
       "      <td>6</td>\n",
       "      <td>species</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>9</td>\n",
       "      <td>32199</td>\n",
       "      <td>species</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   tax_id  parent_tax_id     rank\n",
       "0       1              1  no rank\n",
       "1       2         131567   domain\n",
       "2       6         335928    genus\n",
       "3       7              6  species\n",
       "4       9          32199  species"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "nodes.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ebb7f660",
   "metadata": {},
   "source": [
    "### Traversing the Bifidobacterium sub-tree\n",
    "\n",
    "Since the NCBI taxonomy nodes form a hierarchical tree, I will construct a dictionary mapping parent IDs to their children to efficiently retrieve all descendants of the genus *Bifidobacterium*."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5669d9f4",
   "metadata": {},
   "source": [
    "Now that we cleaned __names__ and __nodes__, we need to identify the tax_id corresponding to the genus *Bifidobacterium*"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "24586630",
   "metadata": {},
   "outputs": [],
   "source": [
    "_ = names[(names['name_txt'].str.contains('bifido',case=False)) & (names['name_class'] == 'scientific name')]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "d5f3df2d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "tax_id                    1678\n",
       "name_txt       Bifidobacterium\n",
       "unique_name                NaN\n",
       "name_class     scientific name\n",
       "Name: 7682, dtype: object"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "_.iloc[0]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a26f2b7f",
   "metadata": {},
   "source": [
    "Bifidobacterium genus taxId is 1678"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "761ee970",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "np.int64(1678)"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "bf_tax_id = _.iloc[0,0]\n",
    "bf_tax_id"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "17bfe226",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Build a mapping from parent tax_id to list of children\n",
    "children = defaultdict(list)\n",
    "for i,row in nodes.iterrows():\n",
    "    parent = row['parent_tax_id']\n",
    "    child = row['tax_id']\n",
    "    children[parent].append(child)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "5065341f",
   "metadata": {},
   "outputs": [],
   "source": [
    "unseen = [bf_tax_id]  # nodes to explore\n",
    "found = set()   # store all found descendant tax_ids\n",
    "\n",
    "while unseen:\n",
    "    actual = unseen.pop()# take the next node to explore\n",
    "\n",
    "    for child in children.get(actual, []):\n",
    "        if child not in found: # skip if already found to avoid duplicates\n",
    "            found.add(child)\n",
    "            unseen.append(child)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "7713efd0",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "903"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(found)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "ac7d6a7f",
   "metadata": {},
   "outputs": [],
   "source": [
    "all_bifido_taxids = nodes.loc[nodes['tax_id'].isin(found),'tax_id']"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1da6464a",
   "metadata": {},
   "source": [
    "### Determining whether the strains possess the hutH gene\n",
    "\n",
    "To address this question, we will use the Biopython library, specifically the `Entrez` module, to retrieve relevant genomic information."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "94c7f305",
   "metadata": {},
   "outputs": [],
   "source": [
    "from Bio import Entrez\n",
    "\n",
    "Entrez.email = \"eliaszorrillagaldon@gmail.com\"\n",
    "\n",
    "# Prepare a list of tax_ids to query NCBI\n",
    "taxids = found\n",
    "\n",
    "results = []\n",
    "\n",
    "\n",
    "# Query the NCBI protein database to check whether each strain has the HutH gene (EC 4.3.1.3)\n",
    "# Using Entrez.esearch, we retrieve the count of matching proteins for each tax_id\n",
    "# If the count is greater than 0, we record that HutH is present\n",
    "\n",
    "\n",
    "for taxid in taxids:\n",
    "    term = f\"txid{taxid}[Organism:exp] AND 4.3.1.3[EC Number]\" \n",
    "\n",
    "    handle = Entrez.esearch(\n",
    "        db=\"protein\",\n",
    "        term=term,\n",
    "        retmax=0\n",
    "    )\n",
    "    record = Entrez.read(handle)\n",
    "    handle.close()\n",
    "\n",
    "    count = int(record[\"Count\"])\n",
    "\n",
    "    results.append({\n",
    "        \"tax_id\": taxid,\n",
    "        \"HutH\": count > 0})\n",
    "\n",
    "    time.sleep(0.34)  # NCBI rate limit\n",
    "\n",
    "# 2. Save the results\n",
    "df_out = pd.DataFrame(results)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "41440a96",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>tax_id</th>\n",
       "      <th>HutH</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>2834432</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>2834433</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2834434</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>2834435</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>2834436</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    tax_id   HutH\n",
       "0  2834432  False\n",
       "1  2834433  False\n",
       "2  2834434  False\n",
       "3  2834435  False\n",
       "4  2834436  False"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_out.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "44aa884f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "903"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(df_out)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f478c0f5",
   "metadata": {},
   "source": [
    "Merge the HutH results with the names DataFrame where `name_class == 'scientific name'`. \n",
    "\n",
    "This adds the strain names to the output bifido_df_names includes both tax_id, HutH presence, and the name."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "478381b5",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>tax_id</th>\n",
       "      <th>HutH</th>\n",
       "      <th>name_txt</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>2834432</td>\n",
       "      <td>False</td>\n",
       "      <td>Bifidobacterium pongonis</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>2834433</td>\n",
       "      <td>False</td>\n",
       "      <td>Bifidobacterium saguinibicoloris</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2834434</td>\n",
       "      <td>False</td>\n",
       "      <td>Bifidobacterium simiiventris</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>2834435</td>\n",
       "      <td>False</td>\n",
       "      <td>Bifidobacterium miconis</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>2834436</td>\n",
       "      <td>False</td>\n",
       "      <td>Bifidobacterium pluvialisilvae</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    tax_id   HutH                          name_txt\n",
       "0  2834432  False          Bifidobacterium pongonis\n",
       "1  2834433  False  Bifidobacterium saguinibicoloris\n",
       "2  2834434  False      Bifidobacterium simiiventris\n",
       "3  2834435  False           Bifidobacterium miconis\n",
       "4  2834436  False    Bifidobacterium pluvialisilvae"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "bifido_df_names = pd.merge(left=df_out,right=names[names['name_class']=='scientific name'][['tax_id','name_txt']],on='tax_id',how='left')\n",
    "bifido_df_names.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ed95dc70",
   "metadata": {},
   "source": [
    "Merge the `bifido_df_names` df with the `nodes` DataFrame to create a dataset containing both tax_ids, huth pressence and their corresponding names and rank."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "8014c987",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>tax_id</th>\n",
       "      <th>HutH</th>\n",
       "      <th>name_txt</th>\n",
       "      <th>rank</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>2834432</td>\n",
       "      <td>False</td>\n",
       "      <td>Bifidobacterium pongonis</td>\n",
       "      <td>species</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>2834433</td>\n",
       "      <td>False</td>\n",
       "      <td>Bifidobacterium saguinibicoloris</td>\n",
       "      <td>species</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2834434</td>\n",
       "      <td>False</td>\n",
       "      <td>Bifidobacterium simiiventris</td>\n",
       "      <td>species</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>2834435</td>\n",
       "      <td>False</td>\n",
       "      <td>Bifidobacterium miconis</td>\n",
       "      <td>species</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>2834436</td>\n",
       "      <td>False</td>\n",
       "      <td>Bifidobacterium pluvialisilvae</td>\n",
       "      <td>species</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>2834437</td>\n",
       "      <td>False</td>\n",
       "      <td>Bifidobacterium miconisargentati</td>\n",
       "      <td>species</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>1798154</td>\n",
       "      <td>False</td>\n",
       "      <td>Bifidobacterium sp. LMM_I9</td>\n",
       "      <td>species</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>1798155</td>\n",
       "      <td>False</td>\n",
       "      <td>Bifidobacterium aerophilum</td>\n",
       "      <td>species</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>1798157</td>\n",
       "      <td>False</td>\n",
       "      <td>Bifidobacterium avesanii</td>\n",
       "      <td>species</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>1798158</td>\n",
       "      <td>False</td>\n",
       "      <td>Bifidobacterium ramosum</td>\n",
       "      <td>species</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    tax_id   HutH                          name_txt     rank\n",
       "0  2834432  False          Bifidobacterium pongonis  species\n",
       "1  2834433  False  Bifidobacterium saguinibicoloris  species\n",
       "2  2834434  False      Bifidobacterium simiiventris  species\n",
       "3  2834435  False           Bifidobacterium miconis  species\n",
       "4  2834436  False    Bifidobacterium pluvialisilvae  species\n",
       "5  2834437  False  Bifidobacterium miconisargentati  species\n",
       "6  1798154  False        Bifidobacterium sp. LMM_I9  species\n",
       "7  1798155  False        Bifidobacterium aerophilum  species\n",
       "8  1798157  False          Bifidobacterium avesanii  species\n",
       "9  1798158  False           Bifidobacterium ramosum  species"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "final_bifido_df = pd.merge(left=bifido_df_names,right=nodes[['tax_id','rank']],on='tax_id',how='left')\n",
    "final_bifido_df.head(10)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "246d01cf",
   "metadata": {},
   "source": [
    "### Save the results\n",
    "\n",
    "Export the Final DataFrame to a `.csv` file for later analysis or sharing."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0105c4e1",
   "metadata": {},
   "outputs": [],
   "source": [
    "folder_path = Path('../data/processed')\n",
    "final_bifido_df.to_csv(folder_path/\"histidine_ammonia_lyase_presence_all_descendants.csv\", index=False)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "dataenv1_new",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
