{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "d8875049",
   "metadata": {},
   "source": [
    "# __Extracting all Bifidobacterium strains__"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "602ca680",
   "metadata": {},
   "source": [
    "### Loading and cleaning the required datasets"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "26c25444",
   "metadata": {},
   "source": [
    "Import required libraries"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "505cc0eb",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import time\n",
    "from pathlib import Path\n",
    "from collections import defaultdict"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b2766b04",
   "metadata": {},
   "source": [
    "We will now load `names.dmp` and `nodes.dmp` as pandas DataFrames from the `new_taxdump_2025-12-01.zip` archive.  \n",
    "This allows us to work with the data in a clean and structured way.\n",
    "\n",
    "> Note: These files are not included in the repository due to their large size.  \n",
    "> You can download the `new_taxdump_2025-12-01` snapshot from the NCBI FTP site:  \n",
    "> [https://ftp.ncbi.nlm.nih.gov/pub/taxonomy/taxdump_archive/](https://ftp.ncbi.nlm.nih.gov/pub/taxonomy/taxdump_archive/)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "3f2d85ed",
   "metadata": {},
   "outputs": [],
   "source": [
    "names_path = Path('../../data/raw/new_taxdump_2025-12-01/names.dmp')\n",
    "names = pd.read_csv(names_path,sep=r\"\\t\\|\\t\",engine='python',header=None)\n",
    "names.columns = ['tax_id','name_txt','unique_name','name_class']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "cb48c0b0",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>tax_id</th>\n",
       "      <th>name_txt</th>\n",
       "      <th>unique_name</th>\n",
       "      <th>name_class</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>all</td>\n",
       "      <td>NaN</td>\n",
       "      <td>synonym\\t|</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>root</td>\n",
       "      <td>NaN</td>\n",
       "      <td>scientific name\\t|</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Bacteria &lt;bacteria&gt;</td>\n",
       "      <td>scientific name\\t|</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>2</td>\n",
       "      <td>bacteria</td>\n",
       "      <td>bacteria &lt;blast name&gt;</td>\n",
       "      <td>blast name\\t|</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>2</td>\n",
       "      <td>bacteria</td>\n",
       "      <td>bacteria &lt;genbank common name&gt;</td>\n",
       "      <td>genbank common name\\t|</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   tax_id  name_txt                     unique_name              name_class\n",
       "0       1       all                             NaN              synonym\\t|\n",
       "1       1      root                             NaN      scientific name\\t|\n",
       "2       2  Bacteria             Bacteria <bacteria>      scientific name\\t|\n",
       "3       2  bacteria           bacteria <blast name>           blast name\\t|\n",
       "4       2  bacteria  bacteria <genbank common name>  genbank common name\\t|"
      ]
     },
     "execution_count": 31,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "names.head() "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "id": "6f6bcbe7",
   "metadata": {},
   "outputs": [],
   "source": [
    "names['name_class'] = names['name_class'].str.strip('\\t|') # Remove unwanted characters from the 'name_class' column"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "id": "38d4ae78",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>tax_id</th>\n",
       "      <th>name_txt</th>\n",
       "      <th>unique_name</th>\n",
       "      <th>name_class</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>all</td>\n",
       "      <td>NaN</td>\n",
       "      <td>synonym</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>root</td>\n",
       "      <td>NaN</td>\n",
       "      <td>scientific name</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Bacteria &lt;bacteria&gt;</td>\n",
       "      <td>scientific name</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   tax_id  name_txt          unique_name       name_class\n",
       "0       1       all                  NaN          synonym\n",
       "1       1      root                  NaN  scientific name\n",
       "2       2  Bacteria  Bacteria <bacteria>  scientific name"
      ]
     },
     "execution_count": 33,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "names.head(3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "id": "2d120c05",
   "metadata": {},
   "outputs": [],
   "source": [
    "nodes_path = Path('../../data/raw/new_taxdump_2025-12-01/nodes.dmp')\n",
    "nodes = pd.read_csv(nodes_path,engine='python',sep=r'\\t\\|\\t',header=None,usecols=[0,1,2],names=['tax_id','parent_tax_id','rank'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "id": "6df01541",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>tax_id</th>\n",
       "      <th>parent_tax_id</th>\n",
       "      <th>rank</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>no rank</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>2</td>\n",
       "      <td>131567</td>\n",
       "      <td>domain</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>6</td>\n",
       "      <td>335928</td>\n",
       "      <td>genus</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>7</td>\n",
       "      <td>6</td>\n",
       "      <td>species</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>9</td>\n",
       "      <td>32199</td>\n",
       "      <td>species</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   tax_id  parent_tax_id     rank\n",
       "0       1              1  no rank\n",
       "1       2         131567   domain\n",
       "2       6         335928    genus\n",
       "3       7              6  species\n",
       "4       9          32199  species"
      ]
     },
     "execution_count": 35,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "nodes.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ebb7f660",
   "metadata": {},
   "source": [
    "### Traversing the Bifidobacterium sub-tree\n",
    "\n",
    "Since the NCBI taxonomy nodes form a hierarchical tree, I will construct a dictionary mapping parent IDs to their children to efficiently retrieve all descendants of the genus *Bifidobacterium*."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5669d9f4",
   "metadata": {},
   "source": [
    "Now that we cleaned __names__ and __nodes__, we need to identify the tax_id corresponding to the genus *Bifidobacterium*"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "id": "24586630",
   "metadata": {},
   "outputs": [],
   "source": [
    "_ = names[(names['name_txt'].str.contains('bifido',case=False)) & (names['name_class'] == 'scientific name')]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "id": "d5f3df2d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "tax_id                    1678\n",
       "name_txt       Bifidobacterium\n",
       "unique_name                NaN\n",
       "name_class     scientific name\n",
       "Name: 7682, dtype: object"
      ]
     },
     "execution_count": 37,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "_.iloc[0]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a26f2b7f",
   "metadata": {},
   "source": [
    "Bifidobacterium genus taxId is 1678"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "id": "761ee970",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "np.int64(1678)"
      ]
     },
     "execution_count": 38,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "bf_tax_id = _.iloc[0,0]\n",
    "bf_tax_id"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "id": "17bfe226",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Build a mapping from parent tax_id to list of children\n",
    "children = defaultdict(list)\n",
    "for i,row in nodes.iterrows():\n",
    "    parent = row['parent_tax_id']\n",
    "    child = row['tax_id']\n",
    "    children[parent].append(child)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "id": "5065341f",
   "metadata": {},
   "outputs": [],
   "source": [
    "unseen = [bf_tax_id]  # nodes to explore\n",
    "found = set()   # store all found descendant tax_ids\n",
    "\n",
    "while unseen:\n",
    "    actual = unseen.pop()# take the next node to explore\n",
    "\n",
    "    for child in children.get(actual, []):\n",
    "        if child not in found: # skip if already found to avoid duplicates\n",
    "            found.add(child)\n",
    "            unseen.append(child)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "id": "de0e2adf",
   "metadata": {},
   "outputs": [],
   "source": [
    "strain_ids = nodes.loc[(nodes['tax_id'].isin(found)) &\n",
    "                       (nodes['rank'].isin(['strain', 'no rank'])),\n",
    "                       'tax_id'\n",
    "                       ]  # Collect all tax_ids whose rank is either 'no rank' or 'strain'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "id": "fb02ac59",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "161"
      ]
     },
     "execution_count": 42,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(strain_ids) # We have 161 taxids "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "aa75c951",
   "metadata": {},
   "source": [
    "Merge the `strain_ids` df with the `names` DataFrame to create a dataset containing both tax_ids and their corresponding names."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "id": "8c4ce879",
   "metadata": {},
   "outputs": [],
   "source": [
    "bifido_strains = pd.merge(left=strain_ids,right=names[names['name_class']=='scientific name'][['tax_id','name_txt']],on='tax_id',how='left') "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "id": "37769347",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>tax_id</th>\n",
       "      <th>name_txt</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>156</th>\n",
       "      <td>1457184</td>\n",
       "      <td>Bifidobacterium longum subsp. longum 72B</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>157</th>\n",
       "      <td>1457185</td>\n",
       "      <td>Bifidobacterium longum subsp. longum EK5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>158</th>\n",
       "      <td>1457186</td>\n",
       "      <td>Bifidobacterium longum subsp. longum EK13</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>159</th>\n",
       "      <td>2608897</td>\n",
       "      <td>unclassified Bifidobacterium</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>160</th>\n",
       "      <td>3436925</td>\n",
       "      <td>Bifidobacterium longum LL6991</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      tax_id                                   name_txt\n",
       "156  1457184   Bifidobacterium longum subsp. longum 72B\n",
       "157  1457185   Bifidobacterium longum subsp. longum EK5\n",
       "158  1457186  Bifidobacterium longum subsp. longum EK13\n",
       "159  2608897               unclassified Bifidobacterium\n",
       "160  3436925              Bifidobacterium longum LL6991"
      ]
     },
     "execution_count": 45,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "bifido_strains.tail()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a91cb0db",
   "metadata": {},
   "source": [
    "### Determining whether the strains possess the hutH gene\n",
    "\n",
    "To address this question, we will use the Biopython library, specifically the `Entrez` module, to retrieve relevant genomic information."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2380cd65",
   "metadata": {},
   "outputs": [],
   "source": [
    "from Bio import Entrez\n",
    "\n",
    "Entrez.email = \"eliaszorrillagaldon@gmail.com\"\n",
    "\n",
    "# Prepare a list of tax_ids to query NCBI\n",
    "taxids = bifido_strains[\"tax_id\"].astype(int).tolist()\n",
    "\n",
    "results = []\n",
    "\n",
    "\n",
    "# Query the NCBI protein database to check whether each strain has the HutH gene (EC 4.3.1.3)\n",
    "# Using Entrez.esearch, we retrieve the count of matching proteins for each tax_id\n",
    "# If the count is greater than 0, we record that HutH is present\n",
    "\n",
    "\n",
    "for taxid in taxids:\n",
    "    term = f\"txid{taxid}[Organism:exp] AND 4.3.1.3[EC Number]\" \n",
    "\n",
    "    handle = Entrez.esearch(\n",
    "        db=\"protein\",\n",
    "        term=term,\n",
    "        retmax=0\n",
    "    )\n",
    "    record = Entrez.read(handle)\n",
    "    handle.close()\n",
    "\n",
    "    count = int(record[\"Count\"])\n",
    "\n",
    "    results.append({\n",
    "        \"tax_id\": taxid,\n",
    "        \"HutH\": count > 0})\n",
    "\n",
    "    time.sleep(0.34)  # NCBI rate limit\n",
    "\n",
    "# 2. Save the results\n",
    "df_out = pd.DataFrame(results)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d298017e",
   "metadata": {},
   "source": [
    "Merge the HutH results with the original bifido_strains DataFrame. \n",
    "\n",
    "This adds the strain names to the final output so final_df includes both tax_id, HutH presence, and the strain name"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9cac6044",
   "metadata": {},
   "outputs": [],
   "source": [
    "final_df = pd.merge(left=df_out,right=bifido_strains,on='tax_id',how='outer')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0b98c066",
   "metadata": {},
   "source": [
    "### Save the results\n",
    "\n",
    "Export the DataFrame containing HutH presence/absence for each strain to a `.csv` file for later analysis or sharing."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "id": "fab95e83",
   "metadata": {},
   "outputs": [],
   "source": [
    "folder_path = Path('../../data/processed')\n",
    "final_df.to_csv(folder_path/\"histidine_ammonia_lyase_presence_01.csv\", index=False)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "dataenv1_new",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
